pyiron/sphinx/base.py

```diff
@@ -1515,6 +1566,107 @@ def write_spins_constraints(self, file_name="spins.in", cwd=None, spins_str=None
             s.logger.debug("No magnetic moments")
 
 
+class Group(dict):
+
+    """
+    Dictionary-like object to store SPHInX inputs.
+
+    Attributes (sub-groups, parameters, & flags) can be set
+    and accessed via dot notation, or as standard dictionary
+    key/values.
+
+    `to_sx_str` converts the Group to the c++-like format
+    expected by SPHInX, for writing files.
+
+    E.g.
+    ```python
+    Group.set("vdwCorrection", {"method": "D2"})
+    Group.set("nEmptyStates", 15)
+    ```
+    """
+
+    def __init__(self, *args, **kw):
+        super(Group, self).__init__(*args, **kw)
+
+    def __setitem__(self, key, value):
+        if isinstance(value, dict):
+            value = Group(value)
+        super(Group, self).__setitem__(key, value)
+
+    def __setattr__(self, key, value):
+        if isinstance(value, dict):
+            value = Group(value)
+        super(Group, self).__setitem__(key, value)
+
+    __getattr__ = dict.get
+    __delattr__ = dict.__delitem__
+
+    def __str__(self):
+        return self.to_sx_str()
+
+    def __repr__(self):
+        return self.to_sx_str()
+
+    def set(self, name, content):
+        self[name] = content
+
+    def set_group(self, name, content=None):
+        if content is None:
+            content = Group()
+        self.set(name, content)
+
+    def set_flag(self, flag, val=True):
+        self.set(flag, val)
+
+    def set_parameter(self, parameter, val):
+        self.set(parameter, val)
+
+    def remove(self, name):
+        del self[name]
+
+    def to_sx_str(self, content="__self__", level=0):
```

Comment:
`to_sx_str` - this might be a function name the Sphinx developers remember, but to make this feature more generic and hopefully be included in other parts of pyiron we should think about a more generic name.

Reply:
Maybe we can write something like `get_C_syntax`, `get_xml_syntax` and `get_python_syntax` in the hdf5-class (or maybe also `print(syntax=XXX)`). For example for

```
job['input/basis/eCut'] = 0.5
```

`job['input'].get_XXX_syntax()` (or `job['input'].print(syntax=XXX)`) returns

```
basis {
    eCut = 0.5;
}
```

```
<basis>
    eCut = 0.5 # I'm not sure if it should be <eCut> 0.5 </eCut>
</basis>
```

and

```
basis:
    eCut = 0.5
```

None of them actually follows the real syntax, but this could make it easier for the user as well to see all the contents of a certain group they're interested in.

Reply:
but isn't the output specific to sx-format? I'd imagine if you reuse this (or whatever @pmrv develops) for other output formats, you'd introduce another to_whatever_str function.

Reply:
> Maybe we can write something like get_C_syntax, get_xml_syntax and get_python_syntax in the hdf5-class (or maybe also print(syntax=XXX))

I like `print(syntax)` a lot. Then we just need if's and elif's for each syntax we support.

Reply:
> but isn't the output specific to sx-format? I'd imagine if you reuse this (or whatever @pmrv develops) for other output formats, you'd introduce another to_whatever_str function.

I guess it'll be independent of @pmrv, because the data has to be stored in hdf5 and we should be able to restore all the input files from there. It's indeed probably an sx-specific format, but for other codes I hope either the user can do small modifications themselves (e.g. via `replace`), or maybe we can add a few arguments which account for the modifications required.

But one way or other, I don't think it's a bad idea to have a `print` function that returns the full content of an hdf5-file (or one of its subgroups), because I sometimes find it frustrating to have to go through each tag to find the data I look for.

Reply:
Agree with @freyso, this is a sx-specific format and should be labelled as such.  I would think for now `str(job.input)` is generic enough in that it works without the user knowing what format to expect.  If we want to add multiple output formats also that feels like a separate project to me. @samwaseda  is right though in that this would ideally more or less independent of my changes.